File: en/codes/go/chapter_backtracking/subset_sum_i_naive.go
// File: subset_sum_i_naive.go
// Created Time: 2023-06-24
// Author: Reanon (793584285@qq.com)

package chapter_backtracking

/* Backtracking algorithm: Subset sum I */
func backtrackSubsetSumINaive(total, target int, state, choices *[]int, res *[][]int) {
	// When the subset sum equals target, record the solution
	if target == total {
		newState := append([]int{}, *state...)
		*res = append(*res, newState)
		return
	}
	// Traverse all choices
	for i := 0; i < len(*choices); i++ {
		// Pruning: if the subset sum exceeds target, skip this choice
		if total+(*choices)[i] > target {
			continue
		}
		// Attempt: make choice, update element sum total
		*state = append(*state, (*choices)[i])
		// Proceed to the next round of selection
		backtrackSubsetSumINaive(total+(*choices)[i], target, state, choices, res)
		// Backtrack: undo choice, restore to previous state
		*state = (*state)[:len(*state)-1]
	}
}

/* Solve subset sum I (including duplicate subsets) */
func subsetSumINaive(nums []int, target int) [][]int {
	state := make([]int, 0) // State (subset)
	total := 0              // Subset sum
	res := make([][]int, 0) // Result list (subset list)
	backtrackSubsetSumINaive(total, target, &state, &nums, &res)
	return res
}

File: en/codes/csharp/chapter_backtracking/subset_sum_i_naive.cs
/**
* File: subset_sum_i_naive.cs
* Created Time: 2023-06-25
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_backtracking;

public class subset_sum_i_naive {
    /* Backtracking algorithm: Subset sum I */
    void Backtrack(List<int> state, int target, int total, int[] choices, List<List<int>> res) {
        // When the subset sum equals target, record the solution
        if (total == target) {
            res.Add(new List<int>(state));
            return;
        }
        // Traverse all choices
        for (int i = 0; i < choices.Length; i++) {
            // Pruning: if the subset sum exceeds target, skip this choice
            if (total + choices[i] > target) {
                continue;
            }
            // Attempt: make choice, update element sum total
            state.Add(choices[i]);
            // Proceed to the next round of selection
            Backtrack(state, target, total + choices[i], choices, res);
            // Backtrack: undo choice, restore to previous state
            state.RemoveAt(state.Count - 1);
        }
    }

    /* Solve subset sum I (including duplicate subsets) */
    List<List<int>> SubsetSumINaive(int[] nums, int target) {
        List<int> state = []; // State (subset)
        int total = 0; // Subset sum
        List<List<int>> res = []; // Result list (subset list)
        Backtrack(state, target, total, nums, res);
        return res;
    }

    [Test]
    public void Test() {
        int[] nums = [3, 4, 5];
        int target = 9;
        List<List<int>> res = SubsetSumINaive(nums, target);
        Console.WriteLine("Input array nums = " + string.Join(", ", nums) + ", target = " + target);
        Console.WriteLine("All subsets with sum equal to " + target + " are res = ");
        foreach (var subset in res) {
            PrintUtil.PrintList(subset);
        }
        Console.WriteLine("Please note that this method outputs results containing duplicate sets");
    }
}
